contracts/interchain-token-service/src/abi.rs

```diff
@@ -0,0 +1,503 @@
+                    )?,
+                })
+            }
+            MessageType::ReceiveFromHub => {
+                let decoded = ReceiveFromHub::abi_decode_params(payload_array, true)
+                    .map_err(|_| MessageError::AbiDecodeFailed)?;
+
+                Self::ReceiveFromHub(types::ReceiveFromHub {
+                    source_chain: String::from_str(env, &decoded.source_chain),
+                    message: Message::abi_decode(
+                        env,
+                        &Bytes::from_slice(env, decoded.message.as_ref()),
+                    )?,
+                })
+            }
+            _ => return Err(MessageError::InvalidMessageType),
+        };
+
+        Ok(message)
+    }
+}
+
+fn soroban_string_to_string(soroban_string: String) -> StdString {
+    let length = soroban_string.len() as usize;
+    let mut bytes = vec![0u8; length];
+    soroban_string.copy_into_slice(&mut bytes);
+    StdString::from_utf8(bytes).expect("Invalid UTF-8 sequence")
+}
+
+fn convert_to_i128(decoded_value: Uint<256, 4>) -> Result<i128, MessageError> {
+    let slice = decoded_value.as_le_slice();
+
+    if !slice[16..].iter().all(|&b| b == 0) {
+        return Err(MessageError::InvalidAmount);
+    }
+
+    let mut truncated = [0u8; 16];
+    truncated.copy_from_slice(&slice[..16]);
+
+    Ok(i128::from_le_bytes(truncated))
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use alloc::vec;
+    use soroban_sdk::{Bytes, BytesN, Env, String};
+    use std::vec::Vec;
+
+    const MAX_I128: i128 = 170141183460469231731687303715884105727 as i128;
+
+    fn bytes_from_hex(env: &Env, hex_string: &str) -> Bytes {
+        let bytes_vec: Vec<u8> = hex::decode(hex_string).unwrap();
+        Bytes::from_slice(env, &bytes_vec)
+    }
+
+    #[test]
+    fn abi_encode_decode() {
+        let env = Env::default();
+
+        let its_messages = vec![
+            types::Message::InterchainTransfer(types::InterchainTransfer {
+                token_id: BytesN::from_array(&env, &[1; 32]),
+                source_address: Bytes::from_slice(&env, &[2; 32]),
+                destination_address: Bytes::from_slice(&env, &[3; 32]),
+                amount: 9_876_543_210_123_456_789,
+                data: Some(Bytes::from_slice(&env, &[4; 32])),
+            }),
+            types::Message::DeployInterchainToken(types::DeployInterchainToken {
+                token_id: BytesN::from_array(&env, &[1; 32]),
+                name: String::from_str(&env, "some_token"),
+                symbol: String::from_str(&env, "TKN"),
+                decimals: 18,
+                minter: Some(Bytes::from_slice(&env, &[1; 32])),
+            }),
+        ];
+
+        for msg in its_messages {
+            let encoded = msg.clone().abi_encode(&env);
+            let decoded = Message::abi_decode(&env, &encoded);
+            assert_eq!(msg, decoded.unwrap());
+        }
+
+        let hub_messages = vec![
```

Comment:
use the existing encode_decode tests if more vectors are needed instead of creating another one